site/cds_rdm/inspire_harvester/writer.py

```diff
@@ -0,0 +1,228 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2025 CERN.
+#
+# CDS-RDM is free software; you can redistribute it and/or modify it under
+# the terms of the MIT License; see LICENSE file for more details.
+
+"""Writer module."""
+import time
+from io import BytesIO
+
+import requests
+from flask import current_app
+from invenio_access.permissions import system_identity
+from invenio_rdm_records.proxies import current_rdm_records_service
+from invenio_records_resources.services.errors import FileKeyNotFoundError
+from invenio_vocabularies.datastreams.errors import WriterError
+from invenio_vocabularies.datastreams.writers import BaseWriter
+
+
+class InspireWriter(BaseWriter):
+    """INSPIRE writer."""
+
+    def __init__(self, *args, **kwargs):
+        """Constructor."""
+        super().__init__(*args, **kwargs)
+
+    def write(self, stream_entry, *args, **kwargs):
+        """Creates or updates the record in CDS."""
+        pass
+
+    def write_many(self, stream_entries, *args, **kwargs):
+        """Creates or updates the record in CDS."""
+        entries = [entry.entry for entry in stream_entries]
+
+        for entry in entries:
+            identifiers = entry["metadata"].get("identifiers", {})
+            inspire_id = next(
+                (
+                    item["identifier"]
+                    for item in identifiers
+                    if item["scheme"] == "inspire"
+                ),
+                None,
+            )
+
+            existing_records = self._get_existing_records(inspire_id)
+            existing_records_hits = existing_records.to_dict()["hits"]["hits"]
+            existing_records_ids = [hit["id"] for hit in existing_records_hits]
+
+            if existing_records.total > 1:
+                raise WriterError(
+                    f"More than 1 record found with INSPIRE id {inspire_id}. CDS records found: {', '.join(existing_records_ids)}"
+                )
+            elif existing_records.total == 1:
+                handle_existing_success = self._handle_existing_records(
+                    entry, existing_records_ids, existing_records_hits, inspire_id
+                )
+                if not handle_existing_success:
+                    continue
+            else:
+                # no existing record in CDS - create and publish a new one
+                record_creation_success = self._create_new_record(inspire_id, entry)
+                if not record_creation_success:
+                    continue
+        return stream_entries
+
+    def _handle_existing_records(
+        self, entry, existing_records_ids, existing_records_hits, inspire_id
+    ):
+        existing_rec_id = existing_records_ids[0]
```

Comment:
you could pass the first result already as an argument, so there would be no need to handle it here